### DISCUSSION TITLE
How to add and use Hangfire dispatcher?

### DISCUSSION THREAD

**Discussion by tsoG2000** [elsa-workflows/elsa-core#1617] [Q&A] [answered]
I guess this is how you add Hangfire to Elsa: ".AddHangfireTemporalActivities(hangfire => hangfire.UseSqlServerStorage("SqlConnectionString"))"  ?

How do you dispatch workflows from HangfireDispatcher?

**Comment by sfmskywalker** [comment, marked as answer]
That overload of `.AddHangfireTemporalActivities` method does indeed add Hangfire services. But there is another overload that does not install Hangfire itself, allowing you to do so yourself.

Adding Hangfire yourself is desirable, especially if you want to leverage Hangfire not just for timer activities, but also for dispatching workflows (to be executed from Hangfire jobs).

To configure the Hangfire provider for workflow dispatch, you can do so like this:

```csharp
services.AddElsa(elsa =>
{
   elsa.UseHangfireDispatcher();
}
```

This requires that you add Hangfire services yourself (either explicitly using Hangfire's extension methods `AddHangfire` and `AddHangfireServer`) or implicitly when using `AddHangfireTemporalActivities()` with the overload that accepts the `configureJobServer` argument.

My recommendation is that you always add Hangfire services yourself using Hangfire's extension methods. Then you're free to use it for just timer activities, the workflow dispatcher, neither or both.

If you wanted to use Hangfire for both timer activities and the dispatcher, your Startup would look like this:

```csharp

services

   // Add Hangfire services.
  .AddHangfire(configuration => configuration
      .UseSimpleAssemblyNameTypeSerializer()
      .UseRecommendedSerializerSettings(settings => settings.ConfigureForNodaTime(DateTimeZoneProviders.Tzdb))
      .UseSqlServerStorage(dbConnectionString, hangfireSqlServerOptions)) // Use SQL Server storage.

   // Server for workflows.
  .AddHangfireServer((sp, options) =>
  {
      options.ConfigureForElsaDispatchers(sp); // Important; the dispatcher uses a dedicated queue ("trigger-workflows-request") for processing workflow requests.
  })
  // Optionally add multiple Hangfire Servers so that you can have different workers for different queues.
  // For example, imagine activities scheduling jobs onto the "heavy-stuff" queue. With this setup, a separate Hangfire server will process that queue, increasing workflow throughput (which are processed from another server with its own set of workers). 
  .AddHangfireServer((sp, options) =>
  {
      options.Queues = new[] { "heavy-stuff" };
  });

// Add Elsa.
services.AddElsa(elsa =>
{
   // Tell Elsa to use Hangfire for dispatching workflows (which means it will schedule Hangfire jobs instead of sending message to a queue via Rebus).
   elsa.UseHangfireDispatcher();

   // Add timer activities using the Hangfire provider.
   elsa.AddHangfireTemporalActivities();

   // Or if you don't want to use Hangfire for timer activities, just use the Quartz provider:
   // elsa.AddQuartzTemporalActivities();
}
```

With above setup, the dispatcher will use Hangfire.

**Reply by b0rv3g4r4** [reply]
I have configured my project like this, but I can't get my workflows to run in hangfire.

The workflow is running but I don't see hangfire running it on its dashboard or on the console

**Comment by dwevedivaibhav** [comment]
Hi @sfmskywalker 
I hope this message finds you well.

I am currently encountering an issue in long running workflow by using hangfire dispatcher. In my AWS ECS deployment, when attempting to execute background jobs for long-running in servers, I've observed that during testing forcibly stopping server or server scale-down, some jobs are reassigned to other servers, while others revert to the old server which is server actually terminated or going to terminate . The issue persists until a new server is started from AWS.

I'm seeking your expertise in resolving this matter. Could you please provide guidance on how we can address and solve this problem?

Your assistance is greatly appreciated!

Thanks in advance.
